default/python/AI/pathfinding.py

```diff
@@ -0,0 +1,139 @@
+    fleet = universe.getFleet(fleet_id)
+    supplied_systems = set(fo.getEmpire().fleetSupplyableSystemIDs)
+
+    if start == INVALID_ID or target == INVALID_ID:
+        warn("Requested path between invalid systems.")
+        return None
+
+    # make sure the target is connected to the start system
+    shortest_possible_path_distance = universe.shortestPathDistance(start, target)
+    if shortest_possible_path_distance == -1:
+        warn("Requested path between disconnected systems, doing nothing.")
+        return None
+
+    # make sure the minimum fuel at target is realistic
+    if minimum_fuel_at_target < 0:
+        error("Requested negative fuel at target.")
+        return None
+
+    if fleet.maxFuel < minimum_fuel_at_target:
+        return None
+
+    start_fuel = fleet.maxFuel if start in supplied_systems else fleet.fuel
+
+    # We have 1 free jump from supplied system into unsupplied systems.
+    # Thus, the target system must be at most maxFuel + 1 jumps away
+    # in order to reach the system under standard conditions.
+    # In some edge cases, we may have more supply here than what the
+    # supply graph suggests. For example, we could have recently refueled
+    # in a system that is now blockaded by an enemy or we have found
+    # the refuel special.
+    target_distance_from_supply = -min(state.get_system_supply(target), 0)
+    if (fleet.maxFuel + 1 < (target_distance_from_supply + minimum_fuel_at_target) and
+            universe.jumpDistance(start, target) > (start_fuel - minimum_fuel_at_target)):
+        # can't possibly reach this system with the required fuel
+        return None
+
+    # initialize data structures
+    path_cache = {}
+    queue = []
+
+    # add starting system to queue
+    heappush(queue, (shortest_possible_path_distance, path_information(distance=0, fuel=start_fuel, path=(start,))))
+
+    while queue:
+        # get next system u with path information
+        (_, path_info) = heappop(queue)
+        current = path_info.path[-1]
+
+        # did we reach the target?
+        if current == target:
+            # do we satisfy fuel constraints?
+            if path_info.fuel < minimum_fuel_at_target:
+                continue
+
+            return path_info
+
+        # add information about how we reached here to the cache
+        path_cache.setdefault(current, []).append(path_info)
+
+        # check if we have enough fuel to travel to neighbors
+        if path_info.fuel < 1:
+            continue
+
+        # add neighboring systems to the queue if the resulting path
+        # is either shorter or offers more fuel than the other paths
+        # which we already found to those systems
+        for neighbor in universe.getImmediateNeighbors(current, empire_id):
+            new_dist = path_info.distance + universe.linearDistance(current, neighbor)
+            new_fuel = (fleet.maxFuel if (neighbor in supplied_systems or current in supplied_systems) else
```

Comment:
`current in supplied_systems` means `fleet.maxFuel - 1`

Reply:
One would intuitively think so. I initially missed this as well.

However, as Dilvish correctly pointed out and as you can easily verify ingame by moving a fleet out of supply, the first jump moving out of supply does not consume any fuel (or is immediately refueled).

Reply:
No, there is no fuel cost to depart from a supplied system.

Reply:
I suggest to add comment about this into code.